src/utils/redstone_tx_sync/client.go

```diff
@@ -0,0 +1,128 @@
+package redstone_tx_sync
+
+import (
+	"context"
+	"encoding/hex"
+	"encoding/json"
+	"math/big"
+	"strings"
+
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/sirupsen/logrus"
+	"github.com/warp-contracts/syncer/src/utils/bundlr"
+	"github.com/warp-contracts/syncer/src/utils/config"
+	"github.com/warp-contracts/syncer/src/utils/logger"
+	"github.com/warp-contracts/syncer/src/utils/model"
+	"github.com/warp-contracts/syncer/src/utils/sequencer"
+	sequencer_types "github.com/warp-contracts/syncer/src/utils/sequencer/types"
+	"gorm.io/gorm"
+)
+
+type Client struct {
+	db              *gorm.DB
+	Log             *logrus.Entry
+	ethClient       *ethclient.Client
+	sequencerClient *sequencer.Client
+}
+
+func NewClient(config *config.Config) (self *Client) {
+	self = new(Client)
+	self.Log = logger.NewSublogger("redstone_tx_sync_client")
+
+	return
+}
+
+func (self *Client) WithEthClient(url EthUrl) *Client {
+	ethClient, err := GetEthClient(self.Log, url)
+	if err != nil {
+		self.Log.WithError(err).Error("Could not get ETH client")
+		return nil
+	}
+	self.ethClient = ethClient
+	return self
+}
+
+func (self *Client) WithDB(db *gorm.DB) *Client {
+	self.db = db
+	return self
+}
+
+func (self *Client) WithSequencerClient(sequencerClient *sequencer.Client) *Client {
+	self.sequencerClient = sequencerClient
+	return self
+}
+
+func (self *Client) GetLastSyncedBlockHeight(ctx context.Context) (lastSyncedHeight int64, err error) {
+	err = self.db.WithContext(ctx).
+		Raw(`SELECT finished_block_height
+		FROM sync_state
+		WHERE name = ?;`, model.SyncedComponentRedstoneTxSyncer).
+		Scan(&lastSyncedHeight).Error
+
+	return
+}
+
+func (self *Client) GetCurrentBlockHeight() (currentBlockHeight int64, currentBlockHash string, err error) {
+	header, err := self.ethClient.HeaderByNumber(context.Background(), nil)
+	if err != nil {
+		return
+	}
+
+	currentBlockHeight = header.Number.Int64()
+	currentBlockHash = header.TxHash.String()
+
+	return
+}
+
+func (self *Client) GetBlockInfo(blockNumber int64) (blockInfo *types.Block, err error) {
+	blockInfo, err = self.ethClient.BlockByNumber(context.Background(), big.NewInt(blockNumber))
+	if err != nil {
+		return
+	}
+	return
+}
+
+func (self *Client) UpdateSyncState(ctx context.Context, blockHeight int64, blockHash string) (err error) {
+	err = self.db.WithContext(ctx).
+		Exec(`UPDATE sync_state
```

Comment:
I'd make updating sync state a separate task performed after syncing and called periodically every 2s. This way if we had 1000 updates in a row there would be only one performed. This would greatly speed up initial sync